File: week4/ObjCRep/ObjCRep/Controllers/Model Controllers/RepresentativeController.m
//
//  RepresentativeController.m
//  ObjCRep
//
//  Created by Eric Lanza on 8/4/18.
//  Copyright © 2018 ETLanza. All rights reserved.
//

#import "RepresentativeController.h"
#import "Representative.h"

@implementation RepresentativeController
{
    NSURL *_baseURL;
    NSMutableArray<Representative*> *_repArray;
}

- (instancetype)init
{
    self = [super init];
    if (self) {
        _baseURL = [[NSURL alloc] initWithString:@"https://whoismyrepresentative.com/getall_reps_bystate.php"];
        _repArray = [NSMutableArray new];
    }
    return self;
}

- (void) searchRepresentativeForState:(NSString *)state completion:(void (^) (NSMutableArray<Representative*> *reps))completion
{
    NSURLQueryItem *stateQuery = [NSURLQueryItem queryItemWithName:@"state" value:state];\
    NSURLQueryItem *outputItem = [NSURLQueryItem queryItemWithName:@"output" value:@"json"];
    
    NSURLComponents *components = [NSURLComponents componentsWithURL:_baseURL resolvingAgainstBaseURL:true];
    components.queryItems = @[stateQuery, outputItem];
    
    NSURL *requestURL = [components URL];
    
    [[NSURLSession.sharedSession dataTaskWithURL:requestURL completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        
        if (error) {
            NSLog(@"Error fecthing data from URL: %@", error);
            return;
        }
        
        if (!data) {
            NSLog(@"Error reading data");
            return;
        }
        
        NSDictionary *topLevelJSON = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&error];
        
        NSArray *jsonArray = topLevelJSON[@"results"];
        
        for (NSDictionary *dictionary in jsonArray) {
            Representative *newRep = [[Representative alloc] initWithName:dictionary[@"name"]];
            [self->_repArray addObject:newRep];
        }
        
        completion(self->_repArray);
        
    }] resume];
}

@end

File: week3/3.4/Representatives/Representatives/Controllers/Model Controllers/RepresentativeController.swift
//
//  RepresentativeController.swift
//  Representatives
//
//  Created by Eric Lanza on 7/19/18.
//  Copyright © 2018 ETLanza. All rights reserved.
//

import Foundation

class RepresentativeController {
    
    static let baseURL = URL(string: "http://whoismyrepresentative.com/getall_reps_bystate.php")
    
    static func searchRepresentatives(forState state: String, completion: @escaping (([Representative]) -> Void)) {
        guard let url = baseURL else { completion([]); return }
        
        var components = URLComponents(url: url, resolvingAgainstBaseURL: true)
        let stateQuery = URLQueryItem(name: "state", value: state.lowercased())
        let jsonQuery = URLQueryItem(name: "output", value: "json")
        components?.queryItems = [stateQuery, jsonQuery]
        guard let requestURL = components?.url else { completion([]); return }
        print(requestURL.absoluteString)
        
        
        let dataTask = URLSession.shared.dataTask(with: requestURL) { (data, _, error) in
            if let error = error {
                print("Error getting representatives: \(error.localizedDescription)")
            }
            guard let data = data,
            let responseDataString = String(data: data, encoding: .ascii),
            let fixedData = responseDataString.data(using: .utf8)
            else { completion([]); return }
            
            let jsonDecoder = JSONDecoder()
            
            do {
                let decodedData = try jsonDecoder.decode([String: [Representative]].self, from: fixedData)
                let repArray = decodedData["results"]
                completion(repArray ?? [])
            } catch {
                print("Error decoding json: \(error.localizedDescription)")
            }
        }
        dataTask.resume()
    }
}
